feat: Defining Greedoid

qawbecrdtey opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

The ~~working~~ old branch for this project is [greedoid-defs](https://github.com/leanprover-community/mathlib4/tree/greedoid-defs). This branch does not contain all features of the ~~working~~ old branch, and is for real contribution to Mathlib.

The current branch have three files `Accessible.lean`, `Exchange.lean` and `Basic.lean`.

Before greedoid is properly defined, it defines two concepts on set systems, namely `ExchangeProperty` in `Exchange.lean` and `AccessibleProperty` and `Accessible` in `Accessible.lean`.

`Greedoid` is a structure of set systems containing empty set, which satisfies an `ExchangeProperty`, and is defined in `Basic.lean`.

- Edit(4de12f4e7a4fb883e23b3447ed648d2bff6ac87d): Removed `Basic.lean` (will be added later), changed pull request title.
- Edit(31f82971f047caa60bc9dc48116cc07dc1d307fc): Restored `Basic.lean`, but removed theorems related to `exchangeProperty` and `accessibleProperty`. Removed theorems are likely to be added in the next commit.
- Edit(9494d7b05f1d2cf517cb7816e0a02c4d23aa4520): Splitted into three files (Accessible/Exchange/Basic). Now `Greedoid` uses `Finset α → Prop` as a feasible set. `ground_set` is present as there are some appearance of minors in the literature.

qawbecrdtey committed:
Initial commit.

qawbecrdtey committed:
cleanup for merge

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Proved 'rank_eq_bases_card', added 'unbot_le_iff'

qawbecrdtey committed:
Added theorem 'bases_singleton'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
proved 'rank_eq_bases_card'

qawbecrdtey committed:
Added some theorems on empty and singleton sets

qawbecrdtey committed:
renamed several theorems

qawbecrdtey committed:
Saving for merge

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Proved 4 theorems

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Reordered theorems.

qawbecrdtey committed:
Added 'induction_on_accessible', 'mem_bases_self_iff', 'rank_of_feasible'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Added

qawbecrdtey committed:
Added 'bases_of_feasible_eq_singleton'

qawbecrdtey committed:
Proved local_submodularity

qawbecrdtey committed:
Added closures

qawbecrdtey committed:
Proved stronger_local_submodularity_left

qawbecrdtey committed:
Proved ssubset_of_feasible_rank

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Working

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Working on 'greedoidRankAxioms_unique_greedoid' and 'rank.toGreedoid'

qawbecrdtey committed:
Working on 'rank_closure_eq_rank_self'

qawbecrdtey committed:
small fix

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Proved several theorems on closure

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Added bases_of_empty, feasible_iff_elem_notin_closure_minus_elem

qawbecrdtey committed:
Added some theorems and 'exists_subset_basis_of_subset_bases'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
proved 'exists_subset_basis_of_subset_bases'

qawbecrdtey committed:
removed unused variables.

qawbecrdtey committed:
Declared 'bases_subset_of_rank_eq_subset'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Removed false lemma, prioved bases_subset_of_rank_eq_of_subset

qawbecrdtey committed:
Proved 'closure_eq_of_subset_adj_closure'

qawbecrdtey committed:
Fixed all errors & Proved all theorems up to now.

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Added more theorems on closure

qawbecrdtey committed:
Added kernel and related stuffs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Added more theorems

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Proved some theorems

qawbecrdtey committed:
Proved closure_kernel_eq_closure

qawbecrdtey committed:
Added two simp lemmas

qawbecrdtey committed:
Added more theorems

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
working..

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Basic properties of rank & closure of greedoids are ready

qawbecrdtey committed:
Added several theorems on basisRank.

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Modified

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Fixed small error due to update

qawbecrdtey committed:
New definition

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Literally stuck. Help.

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Version up

qawbecrdtey committed:
Added 'base_nonempty'

qawbecrdtey committed:
Proved 'exists_feasible_satisfying_basisRank

qawbecrdtey committed:
Many many many fixes

qawbecrdtey committed:
Working...

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Proved 'exists_superset_feasible_satisfying_basisRank'

qawbecrdtey committed:
Working on 'rankFeasible_TFAE'

qawbecrdtey committed:
Proved 'rankFeasible_TFAE'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Working on 'basicRank_union_add_rank_inter_le_basisRank_add_basisRank'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Proved 'basisRank_union_add_rank_inter_le_basisRank_add_basisRank'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Fixed naming

qawbecrdtey committed:
Working on 'rankFeasible_iff_subset_subset_monotoneClosure'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Modified definition

qawbecrdtey committed:
Modified theorem

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Working on 'rankFeasible_iff_subset_subset_monotoneClosure'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Fixes, additions, working on 'rankFeasible_iff_subset_subset_monotoneClosure'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Refactor

qawbecrdtey committed:
Working on 'rankFeasible_iff_subset_subset_monotoneClosure'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Working on 'rankFeasible_iff_subset_subset_monotoneClosure'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
refine

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Fixed errors coming from version update

qawbecrdtey committed:
Added 'mem_base_iff_feasible_and_card_eq_base'

qawbecrdtey committed:
Modified name.

qawbecrdtey committed:
Added some more.

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Added and removed several theorems.

qawbecrdtey committed:
Proved 'rankFeasible_iff_subset_subset_monotoneClosure'

qawbecrdtey committed:
Added 'rankFeasible_iff_monotoneClosure_equal_for_all_basis'

qawbecrdtey committed:
Cleanup

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Added 'closureFeasible'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Replaced deprecated names

qawbecrdtey committed:
Added 'closure_insert_eq_iff_rank_eq'

qawbecrdtey committed:
Added 'rank_le_of_rankFeasible_insert_not_mem_and_kernelClosureOperator_neq'

qawbecrdtey committed:
Working on 'kernelClosureOperator_weak_exchange_property_over_rankFeasible'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Added 'exists_basis_containing_element_if_rank_insert_increases'

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Modified a lot and added several lemmas

qawbecrdtey committed:
Working on kernelClosureOperator_weak_exchange_property_over_rankFeasible

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
working

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Working...

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Merge branch 'master' into greedoid-defs

qawbecrdtey committed:
Fixed termination_by

qawbecrdtey committed:
Reduced to two separate files for initial contribution of greedoid.

qawbecrdtey committed:
Removed Basic.lean for now, and deleted unnecessary imports.

qawbecrdtey committed:
Added copyright header

qawbecrdtey committed:
Attached 'Greedoid.' infront of each name.

qawbecrdtey commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,172 @@
+
+/-- The exchange property of greedoid.
+    Note that the exchange property also hold for matroids. -/
+def Greedoid.ExchangeProperty {α : Type _} [DecidableEq α] (Sys : Finset (Finset α)) :=
```
@ericrbg Thanks for the suggestion!
Using namespace looks like a good suggestion, I'll adopt it in the next commit.

`Greedoid` will have its own type, as shown in the [working branch](https://github.com/leanprover-community/mathlib4/blob/9a42ee5c3eca67135fc38906c9f9b3d32daebc5e/Mathlib/Data/Greedoid/Basic.lean#L176-L180). It was also in the file `Basic.lean`, but I received a suggestion to add the file in the next contributing push. To summarize, the `Greedoid` structure will be there, but not now.

qawbecrdtey replied:
I will, thanks!

qawbecrdtey committed:
Added Greedoid namespace

qawbecrdtey committed:
Added module doc

qawbecrdtey committed:
Fixed typo in doc

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

urkud commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,193 @@
+
+/-- The exchange property of greedoid.
+    Note that the exchange property also hold for matroids. -/
+def ExchangeProperty {α : Type _} [DecidableEq α] (Sys : Finset (Finset α)) :=
+  {s₁ : Finset α} → s₁ ∈ Sys →
+  {s₂ : Finset α} → s₂ ∈ Sys →
+  s₂.card < s₁.card →
+  ∃ x ∈ s₁ \ s₂, insert x s₂ ∈ Sys
```
I'm sorry that your PR sits here for a long time with no reviews.
It seems that most (all?) of Mathlib maintainers don't feel to be experts in this area...

I know nothing about greedoids, so I'm going to leave some purely technical remarks.

In this definition, usage of `{}` binders may lead to surprising results like "Lean eats an argument when you don't expect it". This was definitely the case with Lean 3 and may be the case with Lean 4. I suggest using `⦃⦄` to prevent this.

Also, it's possible to formulate this definition without a `DecidableEq` assumption.

```suggestion
def ExchangeProperty {α : Type*} (Sys : Finset (Finset α)) : Prop :=
  ∀ ⦃s : Finset α⦄, s ∈ Sys → ∀ ⦃t : Finset α⦄, t ∈ Sys → t.card < s.card →
    ∃ x ∈ s, ∃ h : x ∉ t, cons x t h ∈ Sys
```

qawbecrdtey replied:
Thanks for the reply!
Frankly speaking, I'm also not an expert in this field. I am also reading [this book](https://link.springer.com/book/10.1007/978-3-642-58191-5) at my own pace to study more about greedy algorithms (although this led me to a whole new world). So the upcoming definitions and theorems will also follow the order of the book. In other words, I'm very happy to get any technical reviews, thanks a lot!

urkud commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,193 @@
+
+instance {α : Type _} [DecidableEq α] :
+    @DecidablePred (Finset (Finset α)) ExchangeProperty :=
+  fun Sys =>
+    if h : ∃ s₁ ∈ Sys, ∃ s₂ ∈ Sys, s₂.card < s₁.card ∧ ∀ x ∈ s₁ \ s₂, insert x s₂ ∉ Sys
+    then isFalse (fun h' => by
+      let ⟨s₁, hs₁, s₂, hs₂, hs₃, hs₄⟩ := h
+      have ⟨_, ha₁, ha₂⟩ := h' hs₁ hs₂ hs₃
+      exact hs₄ _ ha₁ ha₂)
+    else isTrue (by
+      simp at h
+      intro _ hs₁ _ hs₂ hs
+      have ⟨a, ha⟩ := h _ hs₁ _ hs₂ hs
+      exists a; simp only [mem_sdiff, ha, not_false_eq_true, and_self])
```
Unless you're trying to define a computationally efficient instance, this works:
```suggestion
  fun Sys => by unfold ExchangeProperty; infer_instance
```
If you want to define a computationally efficient instance, then please add a docstring saying so.

urkud commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,193 @@
+      have ⟨a, ha⟩ := h _ hs₁ _ hs₂ hs
+      exists a; simp only [mem_sdiff, ha, not_false_eq_true, and_self])
+
+theorem exchangeProperty_exists_superset_of_card_le {α : Type _} [DecidableEq α]
```
If you name this theorem
```suggestion
theorem ExchangeProperty.exists_superset_of_card_le {α : Type*} [DecidableEq α]
```
then you'll be able to use it with dot notation. Also, please declare some `variable`s and use them. E.g.,

```lean
variable {α : Type*} {Sys : Finset (Finset α)} {s t : Finset α}
```
possibly with `DecidableEq α` unless you redefine properties without decidability assumption.

urkud commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,193 @@
+    {Sys : Finset (Finset α)} (hSys : ExchangeProperty Sys)
+    {s₁ : Finset α} (hs₁ : s₁ ∈ Sys)
+    {s₂ : Finset α} (hs₂ : s₂ ∈ Sys)
+    (hs : s₂.card ≤ s₁.card)
+    {n : ℕ} (hn₁ : n ≤ s₁.card) (hn₂ : s₂.card ≤ n) :
```
`hs` follows from `hn₁` and `hn₂`, so you don't need to assume all three.
```suggestion
    {n : ℕ} (hn₁ : n ≤ s₁.card) (hn₂ : s₂.card ≤ n) :
```

urkud commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,193 @@
+  apply And.intro hs
+  intro h₂
+  apply h
+  exact le_antisymm hn₂ (h₂ ▸ hn₁)
```
Here is a golfed version of the whole theorem (assuming your definition):
```
theorem exchangeProperty_exists_superset_of_card_le {α : Type*} [DecidableEq α]
    {Sys : Finset (Finset α)} (hSys : ExchangeProperty Sys)
    {s₁ : Finset α} (hs₁ : s₁ ∈ Sys)
    {s₂ : Finset α} (hs₂ : s₂ ∈ Sys)
    {n : ℕ} (hn₁ : n ≤ s₁.card) (hn₂ : s₂.card ≤ n) :
    ∃ s ∈ Sys, s₂ ⊆ s ∧ s ⊆ s₁ ∪ s₂ ∧ s.card = n := by
  induction n, hn₂ using Nat.le_induction with
  | base => use s₂; simp [*, subset_union_right]
  | succ n hn ihn =>
    rcases ihn (by omega) with ⟨s, hs, hs₂_sub, hs_sub, rfl⟩
    rcases hSys hs₁ hs hn₁ with ⟨x, hx, hx_mem⟩
    rw [mem_sdiff] at hx
    use insert x s
    simp [*, insert_subset_iff, hs₂_sub.trans]
```

urkud commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,193 @@
+
+/-- The accessible property of greedoid -/
+def AccessibleProperty {α : Type _} [DecidableEq α] (Sys : Finset (Finset α)) : Prop :=
+  {s : Finset α} → s ∈ Sys → s ≠ ∅ → ∃ x ∈ s, s \ {x} ∈ Sys
```
We use `s.Nonempty` instead of `s ≠ ∅`.

semorrison commented:
Currently it seems we can't find a technical reviewer for this one. If Mathlib is going to grow in this direction, we're going to need help from people who are already familiar with this, or adjacent, material. @apnelson1, would you be able to give this a review?

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Fixed several defs and thms, more fixes left.

apnelson1 commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,169 @@
+end ExchangeProperty
+
+/-- The accessible property of greedoid -/
+def AccessibleProperty {α : Type*} [DecidableEq α] (Sys : Finset (Finset α)) : Prop :=
```
I understand the reason for having `Sys` to be a property on `Finset`s, but what is the reason for having `Sys` itself only apply to finitely many sets, and to itself be a set? It looks to me like `feasible : Finset α → Prop` might be a better type for describing a greedoid. If the feasible sets are all finite, then induction should still work easily. 

Certainly for matroids, it was very convenient to have independence have type `Set α → Prop` rather than `Set (Set α)` or `Set (Finset α)`.

qawbecrdtey replied:
I chose to use `Finset` of `Finset`s as used in Mathlib/Combinatorics/SetFamily. I also thought that finiteness of the set family should be explicitly stated. May I ask how `Set α → Prop` has benifit of the implentation while implementing matroids?

qawbecrdtey replied:
@apnelson1 I just figured out the file [`Mathlib.Data.Set.Card`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Card.html) which you have submitted last year, and I also think that using finite `Set` with `ncard` should make lives easier! I'll seriously consider porting this project from `Finset` to `Set`, thank you!

semorrison commented:
@qawbecrdtey, it's helpful if you can mark the conversations with Yury above as "resolved" if you've followed the suggestions there. It makes it easier for other reviews coming in to see what has been dealt with, and what is still active.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Several changes:
1. Removed DecidableEq dependency for Greedoid.
2. Various useful lemmas for accessible/exchange properties will be added to the file later.
3. Predefs.lean -> Basic.lean
4. Added definition of greedoid, and list some trivial properties about it.

qawbecrdtey committed:
Modified Mathlib.lean

qawbecrdtey committed:
Indent?

qawbecrdtey committed:
Reformat.

qawbecrdtey commented on Mathlib/Data/Greedoid/Predefs.lean:
```diff
@@ -0,0 +1,193 @@
+        · rw [ht₄, card_insert_of_not_mem ha₂]
+      exact h ▸ ht₁
+    let ⟨s', hs'₁, hs'₂, hs'₃, hs'₄, hs'₅⟩ :=
+      exchangeProperty_exists_feasible_superset_add_element_feasible hSys hs₁ ht₁
```
I'll apply this in the next commit.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

apnelson1 commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,135 @@
+
+end Greedoid
+
+/-- Greedoid is a nonempty (finite) set system satsifying both accessible and exchange property. -/
```
`satisfying`

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Fixed typo.

qawbecrdtey commented:
There are a few things to review:
1. `Finset (Finset α)` or `Finset α -> Prop` or `Set α -> Prop`: Currently it uses `Finset (Finset α)`, as many lemmas on finite family of finite sets in the Combinatorics directory used it. Still I agree there are some conveniences which follow when the family of sets is implemented as `Set α -> Prop` just like `Matroid`, only if we discard computability. `Finset α -> Prop` is an option between those two.
2. Use of `lemma` instead of `theorem`: I could see a new keyword `lemma` in many other files. I wonder if this can be also applied to this file and afterwards.

semorrison commented:
I've never heard of a greedoid, so this is out of ignorance, but I don't see any reason to mention `Finset` in this file. Switching to `Set a -> Prop` seems a good idea.

apnelson1 commented:
> I've never heard of a greedoid, so this is out of ignorance, but I don't see any reason to mention `Finset` in this file. Switching to `Set a -> Prop` seems a good idea.

I can see the reason for feasible sets being `Finset`s; you are in a setting where everything is finite by nature, and you get some computability consequences (being able to use `cons` for example) but I can't understand why `feasible` should itself be a `Finset`. 

Presumably since with your setup there can only be finitely many feasible sets, you will never be considering a greedoid on an infinite type? If not, then why not just use `Fintype a` where it may be needed, and have `feasible : Finset a -> Prop` in the definition. 'feasible' is an adjective rather than a noun, after all.

qawbecrdtey commented:
Let me state again that I couldn't find any appearance of greedoids elaborating infinite sets. So for now it should be okay to state it as a family of finite sets. Even if we consider infinite greedoid of any type, the generalization of greedoids will probably not be what we were expecting, thus rewriting the library is a must in any case of generalization.

The reason I used `Finset (Finset α)` is simply because I referenced some files in Combinatorics/SetFamily ([1](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Shatter.lean#L31), [2](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Kleitman.lean#L35)) while writing the first draft. This way we can easily know the feasible set of greedoid is also finite. However, the major problem with this approach which I noticed while implementing the draft is that it consumes too much memory. For matroids there is a method of converting a family of bases to a family of independent sets, but the similar thing doesn't work for greedoids. So we should store all the feasible sets, including the basis, if we're not dealing with ranks or closures. This should grow exponentially in most cases, so I also think that `Finset (Finset α)` is not a great idea in this sense. Although it is already implemented, and it was easy to take care of the structure using `Finset` of `Finset`s up to now, I think this should be replaced as it is not really 'computable'.

`Set α -> Prop` should work great as in matroids. As mentioned above we can restrict `α` as a `Fintype`, or we may simply restrict the feasibility to be `True` only if the set is finite. Of course there may be several other options, but all these can be simply managed using `Finset α -> Prop` (and probably assuming `DecidableEq α`?). Also,

>  'feasible' is an adjective rather than a noun, after all.

So maybe I should work on implementing again with family of feasible sets being `Finset α -> Prop`, with possible improvements. I'll bring the update at around mid-August. Thanks!

apnelson1 commented:
> Let me state again that I couldn't find any appearance of greedoids elaborating infinite sets. So for now it should be okay to state it as a family of finite sets. Even if we consider infinite greedoid of any type, the generalization of greedoids will probably not be what we were expecting, thus rewriting the library is a must in any case of generalization.
> 
> The reason I used `Finset (Finset α)` is simply because I referenced some files in Combinatorics/SetFamily ([1](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Shatter.lean#L31), [2](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Kleitman.lean#L35)) while writing the first draft. This way we can easily know the feasible set of greedoid is also finite. However, the major problem with this approach which I noticed while implementing the draft is that it consumes too much memory. For matroids there is a method of converting a family of bases to a family of independent sets, but the similar thing doesn't work for greedoids. So we should store all the feasible sets, including the basis, if we're not dealing with ranks or closures. This should grow exponentially in most cases, so I also think that `Finset (Finset α)` is not a great idea in this sense. Although it is already implemented, and it was easy to take care of the structure using `Finset` of `Finset`s up to now, I think this should be replaced as it is not really 'computable'.
> 
> `Set α -> Prop` should work great as in matroids. As mentioned above we can restrict `α` as a `Fintype`, or we may simply restrict the feasibility to be `True` only if the set is finite. Of course there may be several other options, but all these can be simply managed using `Finset α -> Prop` (and probably assuming `DecidableEq α`?). Also,
> 
> > 'feasible' is an adjective rather than a noun, after all.
> 
> So maybe I should work on implementing again with family of feasible sets being `Finset α -> Prop`, with possible improvements. I'll bring the update at around mid-August. Thanks!



> Let me state again that I couldn't find any appearance of greedoids elaborating infinite sets. So for now it should be okay to state it as a family of finite sets. Even if we consider infinite greedoid of any type, the generalization of greedoids will probably not be what we were expecting, thus rewriting the library is a must in any case of generalization.
> 
> The reason I used `Finset (Finset α)` is simply because I referenced some files in Combinatorics/SetFamily ([1](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Shatter.lean#L31), [2](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Kleitman.lean#L35)) while writing the first draft. This way we can easily know the feasible set of greedoid is also finite. However, the major problem with this approach which I noticed while implementing the draft is that it consumes too much memory. For matroids there is a method of converting a family of bases to a family of independent sets, but the similar thing doesn't work for greedoids. So we should store all the feasible sets, including the basis, if we're not dealing with ranks or closures. This should grow exponentially in most cases, so I also think that `Finset (Finset α)` is not a great idea in this sense. Although it is already implemented, and it was easy to take care of the structure using `Finset` of `Finset`s up to now, I think this should be replaced as it is not really 'computable'.
> 
> `Set α -> Prop` should work great as in matroids. As mentioned above we can restrict `α` as a `Fintype`, or we may simply restrict the feasibility to be `True` only if the set is finite. Of course there may be several other options, but all these can be simply managed using `Finset α -> Prop` (and probably assuming `DecidableEq α`?). Also,
> 
> > 'feasible' is an adjective rather than a noun, after all.
> 
> So maybe I should work on implementing again with family of feasible sets being `Finset α -> Prop`, with possible improvements. I'll bring the update at around mid-August. Thanks!

It feels to me like you are mixing ideas about computation and ideas about reasoning in a way that is causing confusion. What exactly are you hoping to achieve with greedoids in lean, computation-wise?

apnelson1 commented:
> > Let me state again that I couldn't find any appearance of greedoids elaborating infinite sets. So for now it should be okay to state it as a family of finite sets. Even if we consider infinite greedoid of any type, the generalization of greedoids will probably not be what we were expecting, thus rewriting the library is a must in any case of generalization.
> > The reason I used `Finset (Finset α)` is simply because I referenced some files in Combinatorics/SetFamily ([1](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Shatter.lean#L31), [2](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Kleitman.lean#L35)) while writing the first draft. This way we can easily know the feasible set of greedoid is also finite. However, the major problem with this approach which I noticed while implementing the draft is that it consumes too much memory. For matroids there is a method of converting a family of bases to a family of independent sets, but the similar thing doesn't work for greedoids. So we should store all the feasible sets, including the basis, if we're not dealing with ranks or closures. This should grow exponentially in most cases, so I also think that `Finset (Finset α)` is not a great idea in this sense. Although it is already implemented, and it was easy to take care of the structure using `Finset` of `Finset`s up to now, I think this should be replaced as it is not really 'computable'.
> > `Set α -> Prop` should work great as in matroids. As mentioned above we can restrict `α` as a `Fintype`, or we may simply restrict the feasibility to be `True` only if the set is finite. Of course there may be several other options, but all these can be simply managed using `Finset α -> Prop` (and probably assuming `DecidableEq α`?). Also,
> > > 'feasible' is an adjective rather than a noun, after all.
> > 
> > 
> > So maybe I should work on implementing again with family of feasible sets being `Finset α -> Prop`, with possible improvements. I'll bring the update at around mid-August. Thanks!
> 
> > Let me state again that I couldn't find any appearance of greedoids elaborating infinite sets. So for now it should be okay to state it as a family of finite sets. Even if we consider infinite greedoid of any type, the generalization of greedoids will probably not be what we were expecting, thus rewriting the library is a must in any case of generalization.
> > The reason I used `Finset (Finset α)` is simply because I referenced some files in Combinatorics/SetFamily ([1](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Shatter.lean#L31), [2](https://github.com/leanprover-community/mathlib4/blob/7fd56cf8f139d774bfa51062f243c999d34f6903/Mathlib/Combinatorics/SetFamily/Kleitman.lean#L35)) while writing the first draft. This way we can easily know the feasible set of greedoid is also finite. However, the major problem with this approach which I noticed while implementing the draft is that it consumes too much memory. For matroids there is a method of converting a family of bases to a family of independent sets, but the similar thing doesn't work for greedoids. So we should store all the feasible sets, including the basis, if we're not dealing with ranks or closures. This should grow exponentially in most cases, so I also think that `Finset (Finset α)` is not a great idea in this sense. Although it is already implemented, and it was easy to take care of the structure using `Finset` of `Finset`s up to now, I think this should be replaced as it is not really 'computable'.
> > `Set α -> Prop` should work great as in matroids. As mentioned above we can restrict `α` as a `Fintype`, or we may simply restrict the feasibility to be `True` only if the set is finite. Of course there may be several other options, but all these can be simply managed using `Finset α -> Prop` (and probably assuming `DecidableEq α`?). Also,
> > > 'feasible' is an adjective rather than a noun, after all.
> > 
> > 
> > So maybe I should work on implementing again with family of feasible sets being `Finset α -> Prop`, with possible improvements. I'll bring the update at around mid-August. Thanks!
> 
> It feels to me like you are mixing ideas about computation and ideas about reasoning in a way that is causing confusion. What exactly are you hoping to achieve with greedoids in lean, computation-wise?

By the way, am I right in deducing that you are at Kaist? I will be giving a talk about matroid formalization on August 13 - see https://dimag.ibs.re.kr/event/2024-08-13/ . I'd be happy to discuss this in person.

qawbecrdtey commented:
> By the way, am I right in deducing that you are at Kaist? I will be giving a talk about matroid formalization on August 13 - see https://dimag.ibs.re.kr/event/2024-08-13/ . I'd be happy to discuss this in person. 

What a coincidence! I'll be looking forward to meeting you then!

qawbecrdtey committed:
branched out of greedoid-for-contrib.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merged branch master

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
working..

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
small improvements

qawbecrdtey committed:
Finished one function

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Basic Exchange Properties.

qawbecrdtey committed:
Added function nonempty_contains_emptyset'.

qawbecrdtey committed:
Added theorem nonempty_contains_emptyset.

qawbecrdtey committed:
Added induction_on_accessible

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
working...

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
working..

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Finished 'construction_on_accessible'.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Cleanup?

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Removed repeated definitions in Greedoid/Basic.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Added documents.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
working on theorem.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib-with-finset-arrow-prop

qawbecrdtey committed:
working on theorem.

qawbecrdtey committed:
Finished basic building block.

qawbecrdtey committed:
Merge branch 'greedoid-for-contrib-with-finset-arrow-prop' into greedoid-for-contrib

Further work on greedoid-for-contrib will be done on its own branch.

qawbecrdtey committed:
Added imports to Mathlib.lean

qawbecrdtey committed:
Lints.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Removed unnecessary helper theorem from Exchange.lean.

qawbecrdtey committed:
removed unnecessary helper theorem.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
removed name of unused variable.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Removed unnecessary codes.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
cleanup lines. TODO: Fix names if necessary (TODOs).

YaelDillies commented on Mathlib/Data/Greedoid/Accessible.lean:
```diff
@@ -0,0 +1,110 @@
+import Init.Data.Nat.Basic
+
+/-!
+This file contains the definition of `AccessibleProperty` and `Accessible` typeclass,
+along with some main properties of accessible set systems.
+
+# Accessible
```
Title first

YaelDillies commented on Mathlib/Data/Greedoid/Accessible.lean:
```diff
@@ -0,0 +1,110 @@
+
+variable {Sys : Finset α → Prop} [Accessible Sys]
+
+theorem nonempty_contains_emptyset
```
it's callred empty

YaelDillies commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,127 @@
+/-- Greedoid is a nonempty (finite) set system satisfying both accessible and exchange property. -/
+structure Greedoid (α : Type*) where
+  /-- The ground set which every element lies on. -/
+  ground_set : Finset α
```
camelCase

YaelDillies commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,127 @@
+  ground_set : Finset α
+  /-- The feasible set of the greedoid. -/
+  feasible : Finset α → Prop
+  contains_emptyset : feasible ∅
```
```suggestion
  feasible_empty : feasible ∅
```

YaelDillies commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,127 @@
+  /-- The ground set which every element lies on. -/
+  ground_set : Finset α
+  /-- The feasible set of the greedoid. -/
+  feasible : Finset α → Prop
```
```suggestion
  Feasible : Finset α → Prop
```

YaelDillies commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,127 @@
+structure Greedoid (α : Type*) where
+  /-- The ground set which every element lies on. -/
+  ground_set : Finset α
+  /-- The feasible set of the greedoid. -/
```
It's not a set anymore but a pedicate

YaelDillies commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,127 @@
+
+/-- Definition of `Finset` in `Greedoid`.
+    This is often called 'feasible'· -/
+protected def Greedoid.mem (G : Greedoid α) (s : Finset α) := G.feasible s
```
Why not simply use `feasible` directly?

YaelDillies reviewed (commented):
I have strain injuries to both wrists, hence must not type much. Please apologise the poor spelling, vague indications and general conciseness.

Thanks for the PR! Generally, you are not quite following the naming conention. Could you please reviewall your lemma names and see whether they follow  https://leanprover-community.github.io/contribute/naming.html ? @apnelson1 , could you please clarifu what the rekationship to matroids is, and whrtther we need both definitions?

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
quick fix.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
minor fixes (comments, naming, etc) left

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Modified documentation comments.

qawbecrdtey committed:
reduced proof length.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

qawbecrdtey committed:
Added basic theorems on bases/basis.

qawbecrdtey committed:
Removed unused imports.

qawbecrdtey committed:
Added one theorem.

qawbecrdtey committed:
Merge remote-tracking branch 'origin' into greedoid-for-contrib

YaelDillies commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -16,10 +16,15 @@ Greedoid is a set system, i.e., a family of sets, over a finite ground set, whic
 * `Greedoid.Feasible`: The feasible property of the greedoid.
+
 * `Greedoid.Bases`: A bases of a greedoid `G` is a maximal feasible set of `G`.
+
```
Was better without the line breaks!

qawbecrdtey committed:
Removed spaces from doc.

qawbecrdtey committed:
Merge branch 'master' into greedoid-for-contrib

madvorak commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,152 @@
+  exchange_property : Greedoid.ExchangeProperty Feasible
+  subset_ground : ∀ s, Feasible s → s ⊆ ground
+
+section Greedoid
```
What is this `section` for?

madvorak commented on Mathlib/Data/Greedoid/Accessible.lean:
```diff
@@ -0,0 +1,108 @@
+variable {α : Type*}
+
+/-- The accessible property of greedoid. -/
+def AccessibleProperty (Sys : Finset α → Prop) : Prop :=
```
Why do you call it `Sys` instead of `S` here?

madvorak commented on Mathlib/Data/Greedoid/Basic.lean:
```diff
@@ -0,0 +1,152 @@
+  ⟨fun h => by apply eq_of_veq <;> simp [h], fun h => by simp [h]⟩
+
+variable {G : Greedoid α}
+variable {s : Finset α}
```
I personally would write `{s s₁ s₂ : Finset α}` together and leave out the explicit arguments from the `variable` declarations.

kbuzzard commented:
Sadly we can find nobody to review or maintain this theory so I think it's best if we close this PR.
